### PR TITLE
Feat: support spherical harmonic

### PR DESCRIPTION
**Pull request by zhuxudong**
### Please check if the PR fulfills these requirements

- [x] The commit message follows our [guidelines](https://github.com/oasis-engine/engine/blob/main/.github/COMMIT_MESSAGE_CONVENTION.md)
- [ ] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)

### What kind of change does this PR introduce? (Bug fix, feature, docs update, ...)
Feature.
### What is the current behavior? (You can also link to an open issue here)
#217 Use cubemap as diffuse reflection texture.
### What is the new behavior (if this is a feature change)?
Use spherical harmonics as diffuse reflection.
### Does this PR introduce a breaking change? (What changes might users need to make in their application due to this PR?)
Yes.

* Deleted `AmbientLight.diffuseTexture`
* Added `AmbientLight.diffuseSphericalHarmonics`
* Deleted `DiffuseMode.Texture`
* Added `DiffuseMode.SphericalHarmonics`

### Demo:
```typescript
  const sh = new SphericalHarmonics3();
  // Generate spherical harmonics from array or baker 
  // ......
  ambientLight.diffuseMode = DiffuseMode.SphericalHarmonics;
  ambientLight.diffuseSphericalHarmonics = sh;
```

###  From Asset which includes sh array
```typescript
 const array = [0.12,0.3,0.2,...]; // 27 numbers
 const sh = new SphericalHarmonics3();
 sh.setValueByArray(array)；
```

###  From [SH Baker](https://github.com/oasis-engine/engine-baker)
```typescript
  const sh = new SphericalHarmonics3();
  SphericalHarmonics3Baker.fromTextureCubeMap(cubeMap, sh);
```

### Export array from spherical harmonics
```typescript
  const array = [];
  const sh = new SphericalHarmonics3();

  sh.toArray(array);
```

image

### CODE REVIEW

File: packages/math/src/SphericalHarmonics3.ts [hunk 1]
```diff
@@ -0,0 +1,242 @@
...
+  /**  The Y(2, -2) coefficient of the SH3. */
+  y2_2: Color = new Color(0, 0, 0, 0);
+  /**  The Y(2, -1) coefficient of the SH3. */
+  y2_1: Color = new Color(0, 0, 0, 0);
+  /**  The Y(2, 0) coefficient of the SH3. */
+  y20: Color = new Color(0, 0, 0, 0);
+  /**  The Y(2, 1) coefficient of the SH3. */
+  y21: Color = new Color(0, 0, 0, 0);
+  /**  The Y(2, 2) coefficient of the SH3. */
+  y22: Color = new Color(0, 0, 0, 0);
```
Comment (GuoLei1990) [inline packages/math/src/SphericalHarmonics3.ts:68]: SphericalHarmonics3 should not have double data structure.

File: packages/math/src/SphericalHarmonics3.ts [hunk 2]
```diff
@@ -0,0 +1,242 @@
...
+
+/**
+ * Use SH3 to represent irradiance environment maps efficiently, allowing for interactive rendering of diffuse objects under distant illumination.
+ * @remarks
+ * https://graphics.stanford.edu/papers/envmap/envmap.pdf
+ * http://www.ppsloan.org/publications/StupidSH36.pdf
+ * https://google.github.io/filament/Filament.md.html#annex/sphericalharmonics
+ */
+export class SphericalHarmonics3 implements IClone {
+  static fromArray(shArray: number[], out: SphericalHarmonics3): void {
```
Comment (GuoLei1990) [inline packages/math/src/SphericalHarmonics3.ts:13]: You can use setValueByArray likce `Vector3`,`Vector2`.

File: packages/math/src/Color.ts [hunk 3]
```diff
@@ -56,11 +56,13 @@ export class Color implements IClone {
    * @param right - The second color to add
    * @param out - The sum of two colors
    */
-  static add(left: Color, right: Color, out: Color): void {
+  static add(left: Color, right: Color, out: Color): Color {
```
Comment (GuoLei1990) [inline packages/math/src/Color.ts:59]: add return comments.

File: packages/math/src/Color.ts [hunk 4]
```diff
@@ -69,11 +71,13 @@ export class Color implements IClone {
    * @param scale - The amount by which to scale the color
    * @param out - The scaled color
    */
-  static scale(left: Color, s: number, out: Color): void {
+  static scale(left: Color, s: number, out: Color): Color {
```
Comment (GuoLei1990) [inline packages/math/src/Color.ts:74]: add return comments.

File: packages/math/src/SphericalHarmonics3.ts [hunk 5]
```diff
@@ -0,0 +1,242 @@
...
+    this.y20.setValue(0, 0, 0, 0);
+    this.y21.setValue(0, 0, 0, 0);
+    this.y22.setValue(0, 0, 0, 0);
+  }
+
+  /**
+   * Clone the value of this coefficients to an array.
+   * @param out - The array
+   */
+  toArray(out: number[]): void {
```
Comment (GuoLei1990) [inline packages/math/src/SphericalHarmonics3.ts:187]: Add `outOffset` param.

File: packages/math/src/SphericalHarmonics3.ts [hunk 6]
```diff
@@ -0,0 +1,242 @@
...
+
+  private _coefficients: Float32Array = new Float32Array(27);
+
+  /**
+   * Get pre-scaled coefficients used in shader.
+   * @remarks
+   * Convert radiance to irradiance with the A_l which is convoluted by the cosine lobe and pre-scale the basis function.
+   * Reference equation [4,5,6,7,8,9] from https://graphics.stanford.edu/papers/envmap/envmap.pdf
+   */
+  get preScaledCoefficients(): Float32Array {
```
Comment (GuoLei1990) [inline packages/math/src/SphericalHarmonics3.ts:78]: Your comments write `Convert radiance to irradiance` already, so why you named this funtion name 'preScaledCoefficients'. This is very strange.

`convertRadianceToIrradiance` is better.

File: packages/math/src/Color.ts (hunk 3 above)
Comment (zhuxudong) [reply packages/math/src/Color.ts:59]: +1

File: packages/math/src/Color.ts (hunk 4 above)
Comment (zhuxudong) [reply packages/math/src/Color.ts:74]: +1

File: packages/math/src/SphericalHarmonics3.ts (hunk 5 above)
Comment (zhuxudong) [reply packages/math/src/SphericalHarmonics3.ts:187]: +1

File: packages/math/src/SphericalHarmonics3.ts (hunk 2 above)
Comment (zhuxudong) [reply packages/math/src/SphericalHarmonics3.ts:13]: +1

File: packages/core/src/lighting/enums/DiffuseMode.ts [hunk 7]
```diff
@@ -4,7 +4,6 @@
 export enum DiffuseMode {
   /** Solid color mode. */
   SolidColor,
-  /** Texture cube mode. */
-  Texture
-  // SphericalHarmonics
+  /** SH mode */
```
Comment (GuoLei1990) [inline packages/core/src/lighting/enums/DiffuseMode.ts:7]: The comment is too simple.

File: packages/math/src/SphericalHarmonics3.ts (hunk 6 above)
Comment (zhuxudong) [reply packages/math/src/SphericalHarmonics3.ts:78]: +1

File: packages/core/src/lighting/AmbientLight.ts [hunk 8]
```diff
@@ -59,6 +60,23 @@ export class AmbientLight {
...
 
+  /**
+   * Diffuse reflection spherical harmonics 3.
+   * @remarks Effective when diffuse reflection mode is `DiffuseMode.SphericalHarmonics`.
+   */
+  get diffuseSphericalHarmonics(): SphericalHarmonics3 {
+    return this._diffuseSphericalHarmonics;
+  }
+
+  set diffuseSphericalHarmonics(sh: SphericalHarmonics3) {
```
Comment (GuoLei1990) [inline packages/core/src/lighting/AmbientLight.ts:71]: use `value` instead of `sh`.

File: packages/core/src/shaderlib/pbr/ibl_diffuse_frag.glsl [hunk 9]
```diff
@@ -1,21 +1,9 @@
...
-
-    #ifndef PHYSICALLY_CORRECT_LIGHTS
-        lightMapIrradiance *= PI;
-    #endif
-
-    irradiance += lightMapIrradiance;
-
-    RE_IndirectDiffuse( irradiance, geometry, material, reflectedLight );
-
+#ifdef O3_USE_DIFFUSE_ENV
```
Comment (GuoLei1990) [inline packages/core/src/shaderlib/pbr/ibl_diffuse_frag.glsl:1]: Optimiation macro name, 'O3_USE_DIFFUSE_ENV' can't clearly express spherical harmonics

Comment (zhuxudong) [reply packages/core/src/shaderlib/pbr/ibl_diffuse_frag.glsl:1]: +1

File: packages/core/src/lighting/enums/DiffuseMode.ts (hunk 7 above)
Comment (GuoLei1990) [inline packages/core/src/lighting/enums/DiffuseMode.ts]: `SH` is what? Abbreviation is not good for commets.

File: packages/math/src/SphericalHarmonics3.ts [hunk 10]
```diff
@@ -0,0 +1,268 @@
...
+    -0.488603, // -1/2 * Math.sqrt(3 / PI)
+    0.488603, //  1/2 * Math.sqrt(3 / PI)
+    -0.488603, // -1/2 * Math.sqrt(3 / PI)
+
+    1.092548, //  1/2 * Math.sqrt(15 / PI)
+    -1.092548, // -1/2 * Math.sqrt(15 / PI)
+    0.315392, //  1/4 * Math.sqrt(5 / PI)
+    -1.092548, // -1/2 * Math.sqrt(15 / PI)
+    0.546274 //  1/4 * Math.sqrt(15 / PI)
+  ];
```
Comment (GuoLei1990) [inline packages/math/src/SphericalHarmonics3.ts:25]: Only compute once, there is no necessary to write value directly.

File: packages/math/src/SphericalHarmonics3.ts [hunk 11]
```diff
@@ -0,0 +1,268 @@
...
+
+  /**
+   * Convert radiance to irradiance with the A_l which is convoluted by the cosine lobe and pre-scale the basis function.
+   * @remarks
+   * Reference equation [4,5,6,7,8,9] from https://graphics.stanford.edu/papers/envmap/envmap.pdf
+   *
+   * @param out - The array
+   * @returns Pre-scaled array
+   */
+  convertRadianceToIrradiance(out: Float32Array): Float32Array {
```
Comment (GuoLei1990) [inline packages/math/src/SphericalHarmonics3.ts:43]: I think this function a part of  reconstruction in CPU and need another part code in GPU.
It doesn’t make much sense to exist independently and provide to developers.
It makes more sense to provide a complete function to obtain color or irradiance pure CPU！
This function can as a internal function.

File: packages/math/src/SphericalHarmonics3.ts [hunk 12]
```diff
@@ -0,0 +1,268 @@
...
+    0.546274 //  1/4 * Math.sqrt(15 / PI)
+  ];
+
+  private static _convolutionKernel = [
+    3.141593, //  PI
+    2.094395, // (2 * PI) / 3,
+    0.785398 //   PI / 4
+  ];
+
+  private _coefficients: Float32Array = new Float32Array(27);
```
Comment (GuoLei1990) [inline packages/math/src/SphericalHarmonics3.ts:33]: coefficients should public and provide for developers like `Vector3`'s `x`,`y`,`z`.

File: packages/math/src/SphericalHarmonics3.ts [hunk 13]
```diff
@@ -0,0 +1,268 @@
...
+  /**
+   * Add radiance to the SH3 in specified direction.
+   * @remarks
+   * Implements `EvalSHBasis` from [Projection from Cube maps] in http://www.ppsloan.org/publications/StupidSH36.pdf.
+   *
+   * @param color - Radiance color
+   * @param direction - Radiance direction
+   * @param solidAngle - Radiance solid angle, dA / (r^2)
+   */
+  addRadiance(color: Color, direction: Vector3, solidAngle: number): void {
```
Comment (GuoLei1990) [inline packages/math/src/SphericalHarmonics3.ts:98]: The meaning of the parameter is direction light, but your function name is 'addRadiance'.
I think addDirectionLight is better and the third param use 'internsity' is better for debelopers.

**Comment by GuoLei1990** [comment]
I create a PR:https://github.com/zhuxudong/engine/pull/4 to optimization.

File: packages/core/src/lighting/enums/DiffuseMode.ts (hunk 7 above)
Comment (zhuxudong) [reply packages/core/src/lighting/enums/DiffuseMode.ts]: +1

File: packages/math/src/SphericalHarmonics3.ts (hunk 10 above)
Comment (zhuxudong) [reply packages/math/src/SphericalHarmonics3.ts]: +1

File: packages/math/src/SphericalHarmonics3.ts (hunk 11 above)
Comment (zhuxudong) [reply packages/math/src/SphericalHarmonics3.ts]: +1

File: packages/math/src/SphericalHarmonics3.ts (hunk 12 above)
Comment (zhuxudong) [reply packages/math/src/SphericalHarmonics3.ts]: +1

File: packages/math/src/SphericalHarmonics3.ts (hunk 13 above)
Comment (zhuxudong) [reply packages/math/src/SphericalHarmonics3.ts]: +1

**Review by GuoLei1990 — APPROVED** [review]
+1

**Pull request merged**